With a refreshnow and shutdownnow file, wait for an idle indexer before quitting the IDE

tdebaets opened the pull request:
Pull request #5 already improved the behavior when creating a `refreshnow` and `shutdownnow` file at the same time, so that the plug-in waits until the refresh/build is done before quitting the IDE. However, after Eclipse has refreshed a project's files, it may also subsequently update the index with any files that were found to have changed. So the behavior could be improved even further to also take the index into account before quitting the IDE, which is the subject of this pull request.

With the change in this pull request, after noticing that Eclipse is done with refreshing/building, the plug-in will now also wait for Eclipse's indexer to become idle. As soon as the indexer has been idle for 15 seconds, it will quit the IDE (in case a `shutdownnow` file was created alongside the `refreshnow` file).

Note that I had to include the aforementioned change to `MANIFEST.MF` in my pull request, as it turns out that this is required for the `org.eclipse.cdt` imports in `RefreshNowJobListener`.

tdebaets committed:
Wait until the indexer is idle before quitting the IDE

When a refreshnow and shutdownnow file are created together and the refresh triggers Eclipse to update its index, wait until the indexer has been idle for a safe amount of time (currently set to 15 seconds) before quitting.
Also made the logInfo() method of the RefreshNowTask class static so that it can be easily called from other classes.

zorzella commented on plugin/src/com/google/zoodiac/refreshnow/RefreshNowJobListener.java:
```diff
@@ -11,21 +11,83 @@
   
   public RefreshNowJobListener(RefreshNowTask task) {
     this.task = task;
+    // We must already register ourselves as index listener here - if we would
```
FWIW this comment is a little cryptic, particularly outside the context of a code review. I'm going to approve the PR, but it'd be great if this was improved.

tdebaets replied:
I agree that it's cryptic and now that I look at it again, I think that missing the 'indexer idle' event isn't a problem because the `done()` method still checks if the indexer is idle anyway. So I might just remove that comment and move the `addIndexerStateListener()` call to `done()`. I'll push a new commit sometime next week.

tdebaets committed:
Move addIndexerStateListener() call to RefreshNowJobListener.done()

Also removed the comment right above the call because it doesn't apply anymore - missing the 'indexer idle' event isn't a problem because RefreshNowJobListener.done() still checks for an idle indexer anyway.

zorzella commented:
I'm having some trouble after merging this PR, while trying to release a new update site. First, when I hit "Build All" under `site.xml`, it told me:

 ```
Missing required plug-in org.eclipse.cdt.core_0.0.0
```

Fine, ok, I installed the "C/C++ Development Tools".  Then it errored with:

```
An internal error occurred during: "Build Site".
Cannot invoke "org.osgi.framework.Bundle.adapt(java.lang.Class)" because "bundle" is null
```

I figured I'd also install the "C/C++ Development Tools SDK". Once Eclipse restarted, I no longer can use the update site project's `site.xml` to build the update site. I saw an error about (I think) missing a handler for the update site nature.

Do you know what is going on and how I can dig myself out of this?

tdebaets commented:
Once again thanks for merging!

I have no idea what could be the cause of the problems you're experiencing, unfortunately. I have just now tried **Build All** on `site.xml` and nothing appears to happen, so I assume that it's successful? Note that this is still on Eclipse 2019-09 though, so what you're experiencing could have been introduced in one of the newer versions.

If all else fails, you could always check relatively easily if older versions are also affected - every version can still be downloaded from https://www.eclipse.org/downloads/packages/release . I recently had to do the same because version 2019-09 didn't work in Windows Subsystem for Linux. A binary search revealed that 2020-03 is the first version that works, so I started using that (I'm not a fan of the more recent versions as their C/C++ indexer is way too slow).

tdebaets commented:
Any luck with releasing a new update site yet?

I have just tried it myself, this time going through the complete process: updating `Bundle-Version` in the plug-in's `MANIFEST.MF`, updating the version in `feature.xml`, removing the old feature from `site.xml` and adding the new feature, and finally clicking on **Build All**. This was definitely successful, as the new feature/plug-in JARs had been generated, as well as `artifacts.jar` and `content.jar` having been updated.

Note that with OpenJDK 21, I also received an error on **Build All** (`The Security Manager is deprecated and will be removed in a future release`), but this was fixed when running Eclipse in OpenJDK 17.

Since I now have the update site changes locally anyway, do you want me to push these or create a pull request?

zorzella commented:
Got a fresh start with the "Eclipse IDE for Eclipse Committers" and then I added the "CDT" package using the Eclipse Marketplace. Everything then just worked out of the box. I committed the 3.8.3 release to github and also uploaded it to my update site. Let me know if it all works for you.

tdebaets commented:
I've just tried the 3.8.3 release and everything works like a charm, thanks!
